File: reference/conformance/python/sharenet_conformance/link.py
"""ShareNet authenticated links — Python conformance implementation
(R3-003 conformance legs; object defined by R3-001 in the Rust core).

Re-implements the registered handshake + frame derivation with the
stdlib (hashlib/hmac) plus the in-tree pure-Python X25519 and
ChaCha20-Poly1305. Conformance only.
"""

from __future__ import annotations

import hashlib
import hmac

from . import chacha20poly1305 as aead
from . import x25519
from .cbor import CborMap, encode

RESPONDER_SIGN_CONTEXT = b"sharenet-link-auth-v1/responder"
INITIATOR_SIGN_CONTEXT = b"sharenet-link-auth-v1/initiator"
HKDF_INFO = b"sharenet-link-session-v1"
LINK_ID_CONTEXT = b"sharenet-link-id-v1"
FRAME_AAD_CONTEXT = b"sharenet-link-frame-v1"


def sha256(*parts: bytes) -> bytes:
    h = hashlib.sha256()
    for p in parts:
        h.update(p)
    return h.digest()


def build_msg1(e_initiator: bytes) -> bytes:
    return encode(CborMap([(1, 1), (2, e_initiator)]))


def build_msg2_content(
    e_responder: bytes, identity_pk: bytes, identity_created: int, capabilities: bytes | None
) -> bytes:
    identity = CborMap(
        [
            (1, 1),
            (2, identity_pk),
            (3, identity_created),
        ]
    )
    entries = [
        (1, 1),
        (2, e_responder),
        (3, identity),
    ]
    if capabilities is not None:
        entries.append((5, capabilities))
    return encode(CborMap(entries))


def build_msg2(
    e_responder: bytes,
    identity_pk: bytes,
    identity_created: int,
    capabilities: bytes | None,
    signature: bytes,
) -> bytes:
    identity = CborMap(
        [
            (1, 1),
            (2, identity_pk),
            (3, identity_created),
        ]
    )
    entries = [
        (1, 1),
        (2, e_responder),
        (3, identity),
        (4, signature),
    ]
    if capabilities is not None:
        entries.append((5, capabilities))
    return encode(CborMap(entries))


def build_msg3_content(
    identity_pk: bytes, identity_created: int, capabilities: bytes | None
) -> bytes:
    identity = CborMap(
        [
            (1, 1),
            (2, identity_pk),
            (3, identity_created),
        ]
    )
    entries = [
        (1, 1),
        (2, identity),
    ]
    if capabilities is not None:
        entries.append((4, capabilities))
    return encode(CborMap(entries))


def build_msg3(
    identity_pk: bytes,
    identity_created: int,
    capabilities: bytes | None,
    signature: bytes,
) -> bytes:
    identity = CborMap(
        [
            (1, 1),
            (2, identity_pk),
            (3, identity_created),
        ]
    )
    entries = [
        (1, 1),
        (2, identity),
        (3, signature),
    ]
    if capabilities is not None:
        entries.append((4, capabilities))
    return encode(CborMap(entries))


def responder_sign_payload(msg1: bytes, msg2_content: bytes) -> bytes:
    return sha256(RESPONDER_SIGN_CONTEXT, msg1, msg2_content)


def initiator_sign_payload(msg1: bytes, msg2: bytes, msg3_content: bytes) -> bytes:
    return sha256(INITIATOR_SIGN_CONTEXT, msg1, msg2, msg3_content)


def hkdf_sha256(ikm: bytes, salt: bytes, info: bytes, length: int) -> bytes:
    # RFC 5869 (extract + expand) over hmac-sha256
    prk = hmac.new(salt, ikm, hashlib.sha256).digest()
    out = b""
    t = b""
    counter = 1
    while len(out) < length:
        t = hmac.new(prk, t + info + bytes([counter]), hashlib.sha256).digest()
        out += t
        counter += 1
    return out[:length]


def derive_session(shared: bytes, msg1: bytes, msg2: bytes, msg3: bytes):
    salt = sha256(msg1, msg2, msg3)
    okm = hkdf_sha256(shared, salt, HKDF_INFO, 64)
    key_i2r, key_r2i = okm[:32], okm[32:]
    link_id = sha256(LINK_ID_CONTEXT, msg1, msg2, msg3)
    return link_id, key_i2r, key_r2i


def seal_frame(
    key: bytes, link_id: bytes, direction: int, seq: int, payload: bytes
) -> bytes:
    nonce = b"\x00\x00\x00\x00" + seq.to_bytes(8, "big")
    aad = FRAME_AAD_CONTEXT + link_id + bytes([direction]) + seq.to_bytes(8, "big")
    sealed = aead.aead_seal(key, nonce, aad, payload)
    return seq.to_bytes(8, "big") + sealed

File: reference/conformance/typescript/link.ts
/**
 * ShareNet authenticated links — TypeScript conformance implementation
 * (R3-003 conformance legs; object defined by R3-001 in the Rust core).
 *
 * Re-implements the registered handshake + frame rules using node:crypto
 * (X25519 via KeyObjects, HKDF) and the in-tree RFC 8439 AEAD so the
 * committed link vectors reproduce byte-exactly. Conformance only — not
 * production crypto (see the Rust protocol core for that).
 */

import { createHash, createPrivateKey, createPublicKey, diffieHellman, hkdfSync } from "node:crypto";
import { aeadSeal } from "./chacha20poly1305.ts";
import type { Value } from "./cbor.ts";
import { encode } from "./cbor.ts";

/** CBOR map value in the TS value model. */
function m(entries: [Value, Value][]): Value {
  return { t: "map", v: entries };
}
import { nodeIdentityWire } from "./identity.ts";

const X25519_PKCS8_PREFIX = Buffer.from("302e020100300506032b656e04220420", "hex");
const X25519_SPKI_PREFIX = Buffer.from("302a300506032b656e032100", "hex");

export const RESPONDER_SIGN_CONTEXT = "sharenet-link-auth-v1/responder";
export const INITIATOR_SIGN_CONTEXT = "sharenet-link-auth-v1/initiator";
export const HKDF_INFO = Buffer.from("sharenet-link-session-v1");
export const LINK_ID_CONTEXT = "sharenet-link-id-v1";
export const FRAME_AAD_CONTEXT = "sharenet-link-frame-v1";

/** X25519 public key from a (raw) 32-byte scalar. */
export function x25519Public(scalar: Uint8Array): Uint8Array {
  const priv = createPrivateKey({
    key: Buffer.concat([X25519_PKCS8_PREFIX, Buffer.from(scalar)]),
    format: "der",
    type: "pkcs8",
  });
  const spki = createPublicKey(priv).export({ format: "der", type: "spki" }) as Buffer;
  return new Uint8Array(spki.subarray(X25519_SPKI_PREFIX.length));
}

/** X25519 shared secret from a raw scalar + raw peer public key. */
export function x25519Shared(scalar: Uint8Array, peerPublic: Uint8Array): Uint8Array {
  const priv = createPrivateKey({
    key: Buffer.concat([X25519_PKCS8_PREFIX, Buffer.from(scalar)]),
    format: "der",
    type: "pkcs8",
  });
  const pub = createPublicKey({
    key: Buffer.concat([X25519_SPKI_PREFIX, Buffer.from(peerPublic)]),
    format: "der",
    type: "spki",
  });
  return new Uint8Array(diffieHellman({ privateKey: priv, publicKey: pub }));
}

export function sha256(...parts: Uint8Array[]): Uint8Array {
  const h = createHash("sha256");
  for (const p of parts) h.update(p);
  return new Uint8Array(h.digest());
}

export interface NodeIdentityFields {
  publicKey: Uint8Array;
  createdAtUnix: bigint;
  displayName: string | null;
}

/** msg1 wire: {1: 1, 2: e_i}. */
export function buildMsg1(eInitiator: Uint8Array): Uint8Array {
  return encode(
    m([
      [{ t: "int", v: 1n }, { t: "int", v: 1n }],
      [{ t: "int", v: 2n }, { t: "bytes", v: eInitiator }],
    ]),
  );
}

/** msg2 content wire (no signature): {1: 1, 2: e_r, 3: identity, 5?: caps}. */
export function buildMsg2Content(
  eResponder: Uint8Array,
  identity: NodeIdentityFields,
  capabilities: Uint8Array | null,
): Uint8Array {
  const entries: [Value, Value][] = [
    [{ t: "int", v: 1n }, { t: "int", v: 1n }],
    [{ t: "int", v: 2n }, { t: "bytes", v: eResponder }],
    [
      { t: "int", v: 3n },
      nodeIdentityWire(identity.publicKey, identity.createdAtUnix, identity.displayName),
    ],
  ];
  if (capabilities !== null) {
    entries.push([{ t: "int", v: 5n }, { t: "bytes", v: capabilities }]);
  }
  return encode(m(entries));
}

export function buildMsg2(
  eResponder: Uint8Array,
  identity: NodeIdentityFields,
  capabilities: Uint8Array | null,
  signature: Uint8Array,
): Uint8Array {
  const entries: [Value, Value][] = [
    [{ t: "int", v: 1n }, { t: "int", v: 1n }],
    [{ t: "int", v: 2n }, { t: "bytes", v: eResponder }],
    [
      { t: "int", v: 3n },
      nodeIdentityWire(identity.publicKey, identity.createdAtUnix, identity.displayName),
    ],
    [{ t: "int", v: 4n }, { t: "bytes", v: signature }],
  ];
  if (capabilities !== null) {
    entries.push([{ t: "int", v: 5n }, { t: "bytes", v: capabilities }]);
  }
  return encode(m(entries));
}

/** msg3 content wire (no signature): {1: 1, 2: identity, 4?: caps}. */
export function buildMsg3Content(
  identity: NodeIdentityFields,
  capabilities: Uint8Array | null,
): Uint8Array {
  const entries: [Value, Value][] = [
    [{ t: "int", v: 1n }, { t: "int", v: 1n }],
    [
      { t: "int", v: 2n },
      nodeIdentityWire(identity.publicKey, identity.createdAtUnix, identity.displayName),
    ],
  ];
  if (capabilities !== null) {
    entries.push([{ t: "int", v: 4n }, { t: "bytes", v: capabilities }]);
  }
  return encode(m(entries));
}

export function buildMsg3(
  identity: NodeIdentityFields,
  capabilities: Uint8Array | null,
  signature: Uint8Array,
): Uint8Array {
  const entries: [Value, Value][] = [
    [{ t: "int", v: 1n }, { t: "int", v: 1n }],
    [
      { t: "int", v: 2n },
      nodeIdentityWire(identity.publicKey, identity.createdAtUnix, identity.displayName),
    ],
    [{ t: "int", v: 3n }, { t: "bytes", v: signature }],
  ];
  if (capabilities !== null) {
    entries.push([{ t: "int", v: 4n }, { t: "bytes", v: capabilities }]);
  }
  return encode(m(entries));
}

export function responderSignPayload(msg1: Uint8Array, msg2Content: Uint8Array): Uint8Array {
  return sha256(
    new TextEncoder().encode(RESPONDER_SIGN_CONTEXT),
    msg1,
    msg2Content,
  );
}

export function initiatorSignPayload(
  msg1: Uint8Array,
  msg2: Uint8Array,
  msg3Content: Uint8Array,
): Uint8Array {
  return sha256(
    new TextEncoder().encode(INITIATOR_SIGN_CONTEXT),
    msg1,
    msg2,
    msg3Content,
  );
}

export function deriveSession(
  sharedSecret: Uint8Array,
  msg1: Uint8Array,
  msg2: Uint8Array,
  msg3: Uint8Array,
): { linkId: Uint8Array; keyI2R: Uint8Array; keyR2I: Uint8Array } {
  const salt = sha256(msg1, msg2, msg3);
  const okm = new Uint8Array(
    hkdfSync("sha256", Buffer.from(sharedSecret), Buffer.from(salt), HKDF_INFO, 64),
  );
  const keyI2R = okm.slice(0, 32);
  const keyR2I = okm.slice(32, 64);
  const linkId = sha256(
    new TextEncoder().encode(LINK_ID_CONTEXT),
    msg1,
    msg2,
    msg3,
  );
  return { linkId, keyI2R, keyR2I };
}

/** Seal one link frame: seq(8, be) || AEAD(key, nonce, aad, payload). */
export function sealFrame(
  key: Uint8Array,
  linkId: Uint8Array,
  direction: 1 | 2,
  seq: bigint,
  payload: Uint8Array,
): Uint8Array {
  const nonce = new Uint8Array(12);
  new DataView(nonce.buffer).setBigUint64(4, seq, false);
  const aad = new Uint8Array(22 + 32 + 1 + 8);
  aad.set(new TextEncoder().encode(FRAME_AAD_CONTEXT), 0);
  aad.set(linkId, 22);
  aad[54] = direction;
  new DataView(aad.buffer).setBigUint64(55, seq, false);
  const sealed = aeadSeal(key, nonce, aad, payload);
  const out = new Uint8Array(8 + sealed.length);
  new DataView(out.buffer).setBigUint64(0, seq, false);
  out.set(sealed, 8);
  return out;
}
